src/unread/unreadModel.js

```diff
@@ -1,54 +1,187 @@
 /* @flow strict-local */
 import Immutable from 'immutable';
-import { createSelector } from 'reselect';
 
 import type { Action } from '../actionTypes';
 import type {
   UnreadState,
+  UnreadStreamsState,
   UnreadPmsState,
   UnreadHuddlesState,
   UnreadMentionsState,
 } from './unreadModelTypes';
 import type { GlobalState } from '../reduxTypes';
-import type { Selector } from '../types';
-import unreadStreamsReducer from './unreadStreamsReducer';
 import unreadPmsReducer from './unreadPmsReducer';
 import unreadHuddlesReducer from './unreadHuddlesReducer';
 import unreadMentionsReducer from './unreadMentionsReducer';
+import {
+  ACCOUNT_SWITCH,
+  EVENT_MESSAGE_DELETE,
+  EVENT_NEW_MESSAGE,
+  EVENT_UPDATE_MESSAGE_FLAGS,
+  LOGOUT,
+  MESSAGE_FETCH_COMPLETE,
+  REALM_INIT,
+} from '../actionConstants';
+import { getOwnUserId } from '../users/userSelectors';
 
-export const getUnreadStreams: Selector<
-  Immutable.Map<number, Immutable.Map<string, Immutable.List<number>>>,
-> = createSelector(
-  state => state.unread.streams,
-  data => {
-    // prettier-ignore
-    // This is an example of the style-guide rule "Always provide a type
-    // when writing an empty `Immutable` value".  Without the explicit type,
-    // `result` gets inferred as `Immutable.Map<number, empty>`, and then
-    // e.g. the `setIn` call could be completely wrong and the type-checker
-    // wouldn't notice.
-    const result: Immutable.Map<number, Immutable.Map<string, Immutable.List<number>>> =
-      Immutable.Map().asMutable();
-    for (const { stream_id, topic, unread_message_ids } of data) {
-      result.setIn([stream_id, topic], Immutable.List(unread_message_ids));
-    }
-    return result.asImmutable();
-  },
-);
+//
+//
+// Selectors.
+//
+
+export const getUnreadStreams = (state: GlobalState): UnreadStreamsState => state.unread.streams;
 
 export const getUnreadPms = (state: GlobalState): UnreadPmsState => state.unread.pms;
 
 export const getUnreadHuddles = (state: GlobalState): UnreadHuddlesState => state.unread.huddles;
 
 export const getUnreadMentions = (state: GlobalState): UnreadMentionsState => state.unread.mentions;
 
+//
+//
+// Reducer.
+//
+
+const initialStreamsState: UnreadStreamsState = Immutable.Map();
+
+// Like `Immutable.Map#map`, but with the update-only-if-different semantics
+// of `Immutable.Map#update`.  Kept for comparison to `updateAllAndPrune`.
```

Comment:
Yeah, I think the idea is that if you're running through all the items in the data structure, then you're probably doing some transformation where the new value is some entirely different kind of thing from the old value, not where you're taking the old value and just maybe modifying it a bit.

If you are making such a transformation, so that all or nearly all the items will be new, then it's likely more efficient to allocate a single new thing and fill that in from scratch, which is what I expect the implementation of `map` to do (though I haven't checked), than to start from the old and make a bunch of intermediate updates.

There's also the fact that if you have an algorithm that relies on something like `updateAll`, where only a few or potentially no elementsare actually being changed, then you're probably better off finding a variation of your algorithm where you go straight to those elements and don't sweep through all the others looking for them. As we do later in this branch :slightly_smiling_face: